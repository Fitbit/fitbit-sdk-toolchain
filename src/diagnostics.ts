import { relative } from 'path';

import chalk from 'chalk';
import dateformat from 'dateformat';
import indentString from 'indent-string';

interface LineAndCharacter {
  line: number;
  character: number;
}

export interface DiagnosticMessage {
  messageText: string;
  category: DiagnosticCategory;
}

export interface Diagnostic {
  messageText: string | DiagnosticMessage[];
  category: DiagnosticCategory;
  target?: DiagnosticTarget;
  file?: {
    path: string;
    position?:
      | {
          start: { line: number };
        }
      | {
          start: LineAndCharacter;
          end?: LineAndCharacter;
        };
  };
}

export enum DiagnosticCategory {
  Warning = 0,
  Error = 1,
  Suggestion = 2,
  Message = 3,
}

export enum DiagnosticTarget {
  App = 'app',
  Companion = 'companion',
  Settings = 'settings',
}

export type DiagnosticHandler = (diagnostic: Diagnostic) => void;

const consoleColors = {
  [DiagnosticCategory.Warning]: chalk.keyword('orange'),
  [DiagnosticCategory.Error]: chalk.red,
  [DiagnosticCategory.Message]: chalk,
  [DiagnosticCategory.Suggestion]: chalk,
};

const categoryNames = {
  [DiagnosticCategory.Warning]: 'warn',
  [DiagnosticCategory.Error]: 'error',
  [DiagnosticCategory.Message]: 'info',
  [DiagnosticCategory.Suggestion]: 'suggestion',
};

export const logDiagnosticToConsole: DiagnosticHandler = (diagnostic) => {
  const { category, messageText, target } = diagnostic;

  let filePosition = '';
  if (diagnostic.file) {
    const { path, position } = diagnostic.file;
    filePosition += ` ${relative(process.cwd(), path)}`;
    if (position) {
      filePosition += `:${position.start.line + 1}`;
      if ('character' in position.start) {
        filePosition += `,${position.start.character + 1}`;
      }
    }
  }

  const now = new Date();
  const timestamp = dateformat(now, 'HH:MM:ss');

  const colorizer = consoleColors[category];
  const categoryName = categoryNames[category];
  const targetString = target ? target : 'build';
  const message =
    typeof messageText === 'string' ? messageText : messageText[0].messageText;

  console.log(
    colorizer(
      `[${timestamp}][${categoryName}][${targetString}]${filePosition} ${message}`,
    ),
  );

  if (typeof messageText !== 'string') {
    for (let i = 1; i < messageText.length; i += 1) {
      console.log(
        consoleColors[messageText[i].category](
          indentString(messageText[i].messageText, 4),
        ),
      );
    }
  }
};
