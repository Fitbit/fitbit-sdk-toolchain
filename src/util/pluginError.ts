import PluginError from 'plugin-error';

import {
  Diagnostic,
  DiagnosticCategory,
  DiagnosticTarget,
} from '../diagnostics';
import indentString from './indentString';

export function isPluginError(value: unknown): value is PluginError {
  // We can't just do an instanceof check as the error object might be
  // constructed from a different version of PluginError. So we need to
  // employ some heuristics.
  if (!(value instanceof Error)) return false;
  const { showStack, plugin, toString } = value as PluginError;
  return (
    typeof showStack === 'boolean' &&
    typeof plugin === 'string' &&
    toString !== Error.prototype.toString
  );
}

/**
 * Determine heuristically whether the error is likely to be from a
 * user error rather than a bug in the toolchain.
 *
 * @param error caught error object
 */
function isProjectBuildError(error: PluginError): boolean {
  return !error.showStack;
}

function hasEnumerableProp<T, K extends string | number | symbol>(
  obj: { [P in K]?: T },
  prop: K,
): obj is { [P in K]: T } {
  // eslint-disable-next-line no-prototype-builtins
  return obj.propertyIsEnumerable(prop) && obj[prop] !== undefined;
}

const ignoredPluginErrorProps = new Set([
  '__safety',
  '_stack',
  'columnNumber',
  'fileName',
  'lineNumber',
  'message',
  'name',
  'plugin',
  'showProperties',
  'showStack',
  'stack',
  'target',
]);

function hasDiagnosticTarget<T extends PluginError>(
  err: T,
): err is T & { target: DiagnosticTarget } {
  return Object.values(DiagnosticTarget).indexOf((err as any).target) !== -1;
}

function convertToDiagnostic(
  error: PluginError<{ columnNumber?: number }>,
): Diagnostic {
  const diagnostic: Diagnostic = {
    category: DiagnosticCategory.Error,
    messageText: `${error.name}: ${error.message}`,
  };

  if (hasDiagnosticTarget(error)) {
    diagnostic.target = error.target;
  }

  // Our own spin on PluginError.prototype._messageDetails().
  if (error.showProperties) {
    const detailKeys = Object.keys(error).filter(
      (key) => !ignoredPluginErrorProps.has(key),
    );
    if (detailKeys.length) {
      const detailsText = detailKeys
        .map((key) => {
          const value = (error as any)[key];

          if (typeof value === 'string' && value.includes('\n')) {
            return `    ${key}:\n${indentString(value, 6)}\n`;
          }

          return `    ${key}: ${value}\n`;
        })
        .join('');

      diagnostic.messageText = [
        {
          category: DiagnosticCategory.Error,
          messageText: diagnostic.messageText as string,
        },
        {
          category: DiagnosticCategory.Message,
          messageText: 'Details:\n' + detailsText,
        },
      ];
    }
  }

  // Some JS runtimes add `fileName` and `lineNumber` properties to all
  // Error instances. These are the location where the error itself was
  // instantiated, which we do not actually want. We are only interested
  // in the error location **in the project sources**. Luckily the
  // runtime-injected properties are non-enumerable so they are easy to
  // detect.
  if (hasEnumerableProp(error, 'fileName')) {
    diagnostic.file = {
      path: error.fileName,
    };
    if (hasEnumerableProp(error, 'lineNumber')) {
      diagnostic.file.position = {
        start: {
          line: error.lineNumber - 1,
          character: hasEnumerableProp(error, 'columnNumber')
            ? error.columnNumber - 1
            : undefined,
        },
      };
    }
  }
  return diagnostic;
}

const pluginError = {
  convertToDiagnostic,
  isPluginError,
  isProjectBuildError,
};
export default pluginError;
