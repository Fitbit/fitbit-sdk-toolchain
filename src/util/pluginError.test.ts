import PluginError from 'plugin-error';

import pluginError from './pluginError';
import { Diagnostic, DiagnosticMessage } from '../diagnostics';

describe('isPluginError', () => {
    it.each([
        ['with a string message', new PluginError('foo', 'message')],
        ['constructed from an Error', new PluginError('bar', new Error('asdf'))],
        ['with showStack enabled', new PluginError('a', new Error('foo'), { showStack: true })],
    ])('correctly identifies a PluginError %s', (_, error) => {
        expect(pluginError.isPluginError(error)).toBe(true);
    });

    it.each([
        ['undefined', undefined],
        ['null', null],
        ['an object', {}],
        ['an object with no prototype', Object.create(null)],
        ['an array', [1, 2, 3]],
        ['a number', 3],
        ['a string', 'foo'],
        ['a plain Error', new Error('foo')],
        ['a TypeError', new TypeError('nope')],
        [
            'an Error object with the same properties as a PluginError',
            Object.assign(new Error('foo'), {
                showStack: false,
                showProperties: true,
                plugin: 'foo',
            }),
        ],
    ])('correctly identifies %s not to be a PluginError', (_, value) => {
        expect(pluginError.isPluginError(value)).toBe(false);
    });
});

describe('isProjectBuildError', () => {
    it('returns true for PluginError with default options', () => {
        expect(pluginError.isProjectBuildError(new PluginError('foo', 'bar'))).toBe(true);
    });

    it('returns false for PluginError with showStack set', () => {
        expect(pluginError.isProjectBuildError(new PluginError('foo', 'bar', { showStack: true }))).toBe(false);
    });
});

describe('convertToDiagnostic', () => {
    it('deals with PluginErrors constructed from another Error', () =>
        expect(pluginError.convertToDiagnostic(new PluginError('foo', new TypeError('ouch')))).toMatchSnapshot());

    describe('when the error has fileName set', () => {
        it('copies the fileName into the diagnostic', () => {
            const fileName = 'app/foo.js';
            const error = new PluginError('asdf', 'blah', { fileName });
            expect(pluginError.convertToDiagnostic(error)).toHaveProperty('file.path', fileName);
        });
    });

    describe('when the error has only columnNumber set but no fileName', () => {
        it('does not copy any position info', () => {
            const error: PluginError<{ columnNumber?: number }> = new PluginError('foo', 'bar');
            error.columnNumber = 5;
            expect(pluginError.convertToDiagnostic(error)).not.toHaveProperty('file');
        });
    });

    describe('when the error has lineNumber and columnNumber set but no fileName', () => {
        it('does not copy any position info', () => {
            const error: PluginError<{ columnNumber?: number }> = new PluginError('foo', 'bar', { lineNumber: 7 });
            error.columnNumber = 5;
            expect(pluginError.convertToDiagnostic(error)).not.toHaveProperty('file');
        });
    });

    describe('when the error has all the position info set', () => {
        it('converts the position info', () => {
            const fileName = 'companion/bar.js';
            const error: PluginError<{ columnNumber?: number }> = new PluginError('foo', 'bar', { fileName, lineNumber: 7 });
            error.columnNumber = 5;
            expect(pluginError.convertToDiagnostic(error)).toHaveProperty('file', {
                path: fileName,
                position: {
                    // PluginError is 1-based; Diagnostic is 0-based.
                    start: { line: 6, character: 4 },
                },
            });
        });
    });

    it('ignores runtime-inserted error position info', () => {
        const error = new PluginError('blah', 'blah', { fileName: 'foo' });
        Object.defineProperties(error, {
            lineNumber: { value: 5, enumerable: false },
            columnNumber: { value: 12, enumerable: false },
        });
        const diagnostic = pluginError.convertToDiagnostic(error);
        expect(diagnostic).toHaveProperty('file', { path: 'foo' });
    });

    it('ignores runtime-inserted error file name', () => {
        const error = new PluginError('blah', 'blah');
        Object.defineProperties(error, {
            fileName: { value: 'node_modules/dont/care.js', enumerable: false },
        });
        const diagnostic = pluginError.convertToDiagnostic(error);
        expect(diagnostic).not.toHaveProperty('file');
    });

    describe('when there are custom properties', () => {
        // tslint:disable-next-line:no-any
        let error: PluginError<{ [key: string]: any }>;
        let diagnostic: Diagnostic;

        beforeEach(() => {
            // tslint:disable-next-line:no-any
            error = new PluginError('foo', 'message') as any;
            error.customProp = 'custom property';
            error.another = 'one here';
            error.haiku = 'This one has newlines\nembedded in the value\nrefridgerator';
            error.lineNumber = 12;
            error.columnNumber = 3;
            error.fileName = 'path/to/some/file.js';

            diagnostic = pluginError.convertToDiagnostic(error);
        });

        it('constructs a diagnostic', () => expect(diagnostic).toMatchSnapshot());

        it('includes the details in a second part of the message', () => {
            expect(diagnostic).toHaveProperty('messageText', expect.any(Array));
            expect(diagnostic.messageText).toHaveLength(2);
        });

        it('outputs the detail property names and values', () => {
            const details = (diagnostic.messageText as DiagnosticMessage[])[1].messageText;
            expect(details).toMatch(/customProp: custom property/);
            expect(details).toMatch(/another: one here/);
        });

        it.each(['fileName', 'lineNumber', 'columnNumber', 'plugin', 'stack', '_stack'])('does not output %s', prop => {
            const details = (diagnostic.messageText as DiagnosticMessage[])[1].messageText;
            expect(details).not.toMatch(prop);
        });

        it('respects showProperties = false', () => {
            error.showProperties = false;
            diagnostic = pluginError.convertToDiagnostic(error);
            expect(diagnostic).toHaveProperty('messageText', expect.any(String));
        });
    });
});
