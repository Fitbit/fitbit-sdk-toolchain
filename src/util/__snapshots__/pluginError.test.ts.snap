// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToDiagnostic deals with PluginErrors constructed from another Error 1`] = `
Object {
  "category": 1,
  "messageText": "TypeError: ouch",
}
`;

exports[`convertToDiagnostic when there are custom properties constructs a diagnostic 1`] = `
Object {
  "category": 1,
  "file": Object {
    "path": "path/to/some/file.js",
    "position": Object {
      "start": Object {
        "character": 2,
        "line": 11,
      },
    },
  },
  "messageText": Array [
    Object {
      "category": 1,
      "messageText": "Error: message",
    },
    Object {
      "category": 3,
      "messageText": "Details:
    customProp: custom property
    another: one here
    haiku:
      This one has newlines
      embedded in the value
      refridgerator
",
    },
  ],
}
`;
